src/kernel/initrd.zig

```diff
@@ -0,0 +1,673 @@
+    const Error = error{ /// The error for an invalid raw ramdisk when parsing.
+    InvalidRamDisk };
+
+    const Self = @This();
+
+    /// A mapping of opened files so can easily retrieved opened files for reading.
+    opened_files: AutoHashMap(*const vfs.Node, *RamdiskHeader),
+
+    /// The underlying file system
+    fs: *vfs.FileSystem,
+
+    /// The allocator used for allocating memory for opening and reading.
+    allocator: *Allocator,
+
+    /// The list of files in the ram disk. These will be pointers into the raw ramdisk to save on
+    /// allocations.
+    files: []RamdiskHeader,
+
+    /// The root node for the ramdisk file system. This is just a root directory as there is not
+    /// subdirectories.
+    root_node: *vfs.Node,
+
+    /// See vfs.FileSystem.instance
+    instance: usize,
+
+    /// See vfs.FileSystem.getRootNode
+    fn getRootNode(fs: *const vfs.FileSystem) *const vfs.DirNode {
+        var self = @fieldParentPtr(RamdiskFS, "instance", fs.instance);
+        return &self.root_node.Dir;
+    }
+
+    /// See vfs.FileSystem.close
+    fn close(fs: *const vfs.FileSystem, node: *const vfs.FileNode) void {
+        var self = @fieldParentPtr(RamdiskFS, "instance", fs.instance);
+        const cast_node = @ptrCast(*const vfs.Node, node);
+        // As close can't error, if provided with a invalid Node that isn't opened or try to close
+        // the same file twice, will just do nothing.
+        if (self.opened_files.contains(cast_node)) {
+            _ = self.opened_files.remove(cast_node);
+            self.allocator.destroy(node);
+        }
+    }
+
+    /// See vfs.FileSystem.read
+    fn read(fs: *const vfs.FileSystem, node: *const vfs.FileNode, len: usize) (Allocator.Error || vfs.Error)![]u8 {
+        var self = @fieldParentPtr(RamdiskFS, "instance", fs.instance);
+        const cast_node = @ptrCast(*const vfs.Node, node);
+        const file_header = self.opened_files.get(cast_node) orelse return vfs.Error.NotOpened;
+        const length = std.math.min(len, file_header.content.len);
+        const buff = try self.allocator.alloc(u8, length);
+        std.mem.copy(u8, buff, file_header.content[0..length]);
+        return buff;
+    }
+
+    /// See vfs.FileSystem.write
+    fn write(fs: *const vfs.FileSystem, node: *const vfs.FileNode, bytes: []const u8) (Allocator.Error || vfs.Error)!void {}
+
+    /// See vfs.FileSystem.open
+    fn open(fs: *const vfs.FileSystem, dir: *const vfs.DirNode, name: []const u8, flags: vfs.OpenFlags) (Allocator.Error || vfs.Error)!*vfs.Node {
+        var self = @fieldParentPtr(RamdiskFS, "instance", fs.instance);
+        switch (flags) {
+            .CREATE_DIR, .CREATE_FILE => return vfs.Error.InvalidFlags,
+            .NO_CREATION => {
+                for (self.files) |*file| {
+                    if (std.mem.eql(u8, file.name, name)) {
+                        // Opening 2 files of the same name, will create 2 different Nodes
```

Comment:
I wanted to create the *Nodes on the fly so not, this would save memory but would be slower. Having all the *Nodes created at init would increase memory, but would be faster. I don't mind having 2 different nodes for the same file. I can iterate through the hash map and search for any node that is opened (I did have this) but this would make opening even slower. What are your thoughts on this?

Reply:
To be honest, at the moment I can't think of any downsides to have two nodes per file (other than the extra memory allocated) so I'm gonna leave it up to you :)

Reply:
I guess as the initrd will only be used once and load all files in one go, I could create all the Nodes and add a `free` or `deinit` to clear the initrd and reuse the memory after everything is loaded. This will also free the raw ramdisk so have all the memories.

Reply:
That sounds good :)